contrib/implements/internal/implements/gosrc.go

```diff
@@ -109,6 +105,29 @@ func isPublic(gfunc *goFunction) bool {
 	return ast.IsExported(gfunc.shortName) && ast.IsExported(gfunc.fullName)
 }
 
+func isPreviewFile(file *ast.File) bool {
+	const previewTag = "ceph_preview"
+	if file.Doc == nil {
+		return false
+	}
+	for _, c := range file.Doc.List {
+		if c == nil {
+			continue
+		}
+		if expr, err := constraint.Parse(c.Text); err == nil {
+			containsPreviewTag := false
+			builtWithoutPreviewTag := expr.Eval(func(tag string) bool {
+				if tag == previewTag {
+					containsPreviewTag = true
+				}
+				return false
+			})
+			return containsPreviewTag && !builtWithoutPreviewTag
```

Comment:
Can builtWithoutPreviewTag ever be true if the `ok` callback above always return false?  If so might it not be simpler to discard the results of `expr.Eval` or am I overlooking something?

Reply:
It could, if someone someone writes `!ceph_preview`. That's anyway just heuristic, because I anyway have the value. Actually I don't think this whole logic is reliable. For example `another_tag && ceph_preview` probably doesn't work, because it will never evaluate the second expression, so it will not "see" the tag. I will rewrite this.